src/VisualStudio/Core/Impl/CodeModel/ProjectCodeModelFactory.cs

```diff
@@ -32,15 +38,91 @@ internal sealed class ProjectCodeModelFactory : IProjectCodeModelFactory, IDispo
+
+        private readonly IForegroundNotificationService _notificationService;
+        private readonly IAsynchronousOperationListener _listener;
+        private readonly AsyncBatchingWorkQueue<DocumentId> _documentsToFireEventsFor;
+
         [ImportingConstructor]
         [SuppressMessage("RoslynDiagnosticsReliability", "RS0033:Importing constructor should be [Obsolete]", Justification = "Used in test code: https://github.com/dotnet/roslyn/issues/42814")]
-        public ProjectCodeModelFactory(VisualStudioWorkspace visualStudioWorkspace, [Import(typeof(SVsServiceProvider))] IServiceProvider serviceProvider, IThreadingContext threadingContext)
+        public ProjectCodeModelFactory(
+            VisualStudioWorkspace visualStudioWorkspace,
+            [Import(typeof(SVsServiceProvider))] IServiceProvider serviceProvider,
+            IThreadingContext threadingContext,
+            IForegroundNotificationService notificationService,
+            IAsynchronousOperationListenerProvider listenerProvider)
         {
             _visualStudioWorkspace = visualStudioWorkspace;
             _serviceProvider = serviceProvider;
             _threadingContext = threadingContext;
             _deferredCleanupTasks = new JoinableTaskCollection(threadingContext.JoinableTaskContext);
             _deferredCleanupTasks.DisplayName = nameof(ProjectCodeModelFactory) + "." + nameof(_deferredCleanupTasks);
+
+            _notificationService = notificationService;
+            _listener = listenerProvider.GetListener(FeatureAttribute.CodeModel);
+
+            // Queue up notifications we hear about docs changing.  that way we don't have to fire events multiple times
+            // for the same documents.  Once enough time has passed, take the documents that were changed and run
+            // through them, firing their latest events.
+            _documentsToFireEventsFor = new AsyncBatchingWorkQueue<DocumentId>(
+                TimeSpan.FromMilliseconds(visualStudioWorkspace.Options.GetOption(InternalSolutionCrawlerOptions.AllFilesWorkerBackOffTimeSpanInMS)),
+                ProcessNextDocumentBatchAsync,
+                // We only care about unique doc-ids, so pass in this comparer to collapse streams of changes for a
+                // single document down to one notification.
+                EqualityComparer<DocumentId>.Default,
+                _listener,
+                _cancellationTokenSource.Token);
+
+            _visualStudioWorkspace.WorkspaceChanged += OnWorkspaceChanged;
+        }
+
+        private System.Threading.Tasks.Task ProcessNextDocumentBatchAsync(
+            ImmutableArray<DocumentId> documentIds, CancellationToken cancellationToken)
+        {
+            foreach (var documentId in documentIds)
+            {
+                // Now, enqueue foreground work to actually process these documents in a serialized and incremental
```

Comment:
❔ Does this operation still show up as pending work in the task status center?

Reply:
This uses the same threading notification mecanism as before (i.e. IAsyncOperationListener), i haven't hcanged that.  Note: i would not expected these to show up in TSC.  For one, there's no mechanism here to present a viable user visible string to describe what is going on.

Reply:
> This uses the same threading notification mecanism as before (i.e. IAsyncOperationListener)

IAsyncOperationListener is a NOP in production. It only exists for test purposes.

> i would not expected these to show up in TSC. For one, there's no mechanism here to present a viable user visible string to describe what is going on.

Today, these do show up because the solution crawler reports progress of incremental analyzer execution in a generalized manner. We should preserve or clone this aspect of the implementation so the value it provides is not lost.

Reply:
> Today, these do show up because the solution crawler reports progress of incremental analyzer execution in a generalized manner. We should preserve or clone this aspect of the implementation so the value it provides is not lost.

1. we'll still get the work notifications from incremental analyzer as it walks the system.
2. I'm really loath to do that for all these one-off cases.  i.e. we have a *lot* of systems which aren't using incremental anlayzer, but have work queues of things to do.  Having anyone who batches and does work need to somehow tie into that doesn't seem useful to me.  Esp. as this isn't a feature which results in a user visible impact.  i.e. i think it's fine to say "we're computing code actions" in that list because that can help users understand why their code actions haven't appeared yet.  Having user notifications for what it effectively a legacy internal symbolic change API just doesn't seem useful.  We would basically be saying "computing code model events for clients listening for them".  That's not valuable information to me as a customer as that doesn't have any meaning for my experience.